Fix decomp behaviour in export training IR

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134657
* #135080
* __->__ #134801

Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

tugsbayasgalan committed:
Fix decomp behaviour in export training IR

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behaviour in export training IR"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

tugsbayasgalan committed:
Update on "Fix decomp behaviour in export training IR"


Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

[ghstack-poisoned]

pianpwk commented on test/export/test_export.py:
```diff
@@ -1467,37 +1465,39 @@ def forward(self, p_conv_weight, p_conv_bias, p_conv1d_weight, p_conv1d_bias, b_
         self.assertExpectedInline(
             str(ep_has_convd.graph_module.code).strip(),
             """\
 def forward(self, p_conv_weight, p_conv_bias, p_conv1d_weight, p_conv1d_bias, b_linear_weight, b_linear_bias, x, y):
```
lol ok this test case makes much more sense now

pianpwk commented on test/export/test_export.py:
```diff
@@ -1727,9 +1727,9 @@ def forward(self, x):
-    sum_1 = torch.ops.aten.sum.default(add)
+    permute = torch.ops.aten.permute.default(p_linear_weight, [1, 0]);  p_linear_weight = None
+    addmm = torch.ops.aten.addmm.default(p_linear_bias, x, permute);  p_linear_bias = x = permute = None
+    sum_1 = torch.ops.aten.sum.dim_IntList(add, [])
```
noob question: why the change here?

pianpwk commented on torch/_export/utils.py:
```diff
@@ -661,15 +876,16 @@ def _detect_fake_mode_from_gm(
-                fake_val = node.meta["val"]
-            if fake_val is not None and isinstance(fake_val, torch.Tensor):
-                fake_vals.append(fake_val)
+
```
hm could you explain this change?

pianpwk replied:
I think this makes sense, but could you explain what prompted the change? The previous implementation seemed equivalent, and only required one loop right?

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1336,6 +1309,13 @@ def _strict_export_lower_to_aten_ir(
 
     _normalize_nn_module_stack(gm_torch_level, type(mod))
 
+    params_buffers_to_node_meta = _collect_param_buffer_metadata(gm_torch_level)
```
nice

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1364,26 +1344,9 @@ def _strict_export_lower_to_aten_ir(
-                        node.meta[k] = v
+    _populate_param_buffer_metadata_to_new_gm(
+        params_buffers_to_node_meta, gm, export_graph_signature
+    )
```
nice

pianpwk commented on torch/_export/utils.py:
```diff
@@ -42,6 +45,175 @@
 }
 
 
+def _collect_and_set_constant_attrs(
```
nit: could you help add some context in the docstrings as to when/where these functions are used? e.g. in strict-mode, after dynamo export. Just makes it easier to parse through this file when a lot's going on.

pianpwk commented on torch/export/exported_program.py:
```diff
@@ -271,87 +280,14 @@ def _override_decomp_aten_to_variants():
         yield
 
 
-def _rename_without_collisions(
```
thanks, this was the wrong place for this in the 1st place

pianpwk commented on torch/export/exported_program.py:
```diff
@@ -655,6 +545,31 @@ def update_arg(old_arg, new_ph):
     return gm, new_graph_signature
 
 
+def _remove_unneccessary_copy_op_pass(
```
noob question: why is this op special?

tugsbayasgalan committed:
Update on "Fix decomp behaviour in export training IR"


Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behaviour in export training IR"


Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behaviour in export training IR"


Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

avikchaudhuri reviewed (approved):
lgtm after discussing offline

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).
